src/lib.rs

```diff
@@ -385,6 +492,74 @@ impl<T: ?Sized> Deref for RwLockReadGuard<'_, T> {
     }
 }
 
+/// A guard that releases the upgradable read lock when dropped.
+pub struct RwLockUpgradableReadGuard<'a, T: ?Sized>(&'a RwLock<T>, Option<MutexGuard<'a, ()>>);
+
+unsafe impl<T: Send + ?Sized> Send for RwLockUpgradableReadGuard<'_, T> {}
+unsafe impl<T: Sync + ?Sized> Sync for RwLockUpgradableReadGuard<'_, T> {}
+
+impl<'a, T: ?Sized> RwLockUpgradableReadGuard<'a, T> {
+    /// Upgrades a upgradable read lock to a writer lock
+    pub async fn upgrade(mut self) -> RwLockWriteGuard<'a, T> {
```

Comment:
Let's change this to:
```suggestion
    pub async fn upgrade(this: Self) -> RwLockWriteGuard<'a, T> {
```

See `parking_lot`'s signature of the equivalent function: https://docs.rs/lock_api/0.4.0/lock_api/struct.RwLockUpgradableReadGuard.html#method.upgrade

The reason for that is we don't want `.upgrade()` on `RwLockUpgradableReadGuard` to clash with `.upgrade()` on the inner type `T`, if it has a method with the same name.

Generally speaking, this rule applies to all types implementing `Deref` (like `RwLockUpgradableReadGuard` does).